### DISCUSSION TITLE
Can I use SQLite math extension in GRDB?

### DISCUSSION THREAD

**Discussion by orkhanalizada2020** [groue/GRDB.swift#1551] [Q&A]
I want to use SQLite math extension in GRDB. Is it possible? If yes, any hint on how can I achieve that?
Thanks in advance!

**Comment by groue** [comment]
Hello @orkhanalizada2020,

I suppose you're talking about the [Built-In Mathematical SQL Functions](https://www.sqlite.org/lang_mathfunc.html).

If SQLite is built with support for those functions, then you can use them, as all SQL functions.

You can use raw SQL (see [Embedding SQL in Query Interface Requests](https://github.com/groue/GRDB.swift/blob/master/README.md#embedding-sql-in-query-interface-requests)):

```swift
// SELECT sin(angle), pow(width, 2) FROM shape
let request = Shape.select(sql: "sin(angle), pow(width, 2)")
```

You can also define Swift versions on the SQL functions needed by your app, as below:

```swift
func sin(_ value: some SQLSpecificExpressible) -> SQLExpression {
    SQL("sin(\(value))").sqlExpression
}

func pow(_ x: some SQLSpecificExpressible, _ y: some SQLExpressible) -> SQLExpression {
    SQL("pow(\(x), \(y))").sqlExpression
}

// SELECT sin(angle), pow(width, 2) FROM shape
let request = Shape.select(sin(Column("angle")), pow(Column("width"), 2))
```

Take care, as in the sample code, to use `SQLSpecificExpressible` for the type of parameters that should accept SQL expressions but no regular Swift numbers, and `SQLExpressible` for parameters that should accept both:

```swift
sin(2) // The regular Swift sine function, not an SQL expression
sin(Column("angle")) // An SQL expression
pow(Column("width"), 2) // An SQL expression
```
